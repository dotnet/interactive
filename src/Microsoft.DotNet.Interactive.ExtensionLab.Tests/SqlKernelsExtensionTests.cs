// // Copyright (c) .NET Foundation and contributors. All rights reserved.
// // Licensed under the MIT license. See LICENSE file in the project root for full license information.


using System;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using Assent;
using Microsoft.Data.SqlClient;
using Microsoft.DotNet.Interactive.CSharp;
using Microsoft.DotNet.Interactive.Formatting;
using Xunit;
using Xunit.Abstractions;

namespace Microsoft.DotNet.Interactive.ExtensionLab.Tests
{
    public class SqlKernelsExtensionTests : IDisposable
    {
        private readonly ITestOutputHelper _output;
        private readonly Configuration _configuration;

        public SqlKernelsExtensionTests(ITestOutputHelper output)
        {
            _output = output;
            _configuration = new Configuration()
                .SetInteractive(Debugger.IsAttached)
                .UsingExtension("json");
        }


        [Fact(Skip = "Requires database")]
        public async Task can_generate_tabular_json_from_database_table_result()
        {
            using var kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
                new KeyValueStoreKernel()
            };

            var extension = new SqlKernelsExtension();

            await extension.OnLoadAsync(kernel);

            var connectionString = "Persist Security Info=False; Integrated Security=true; Initial Catalog=AdventureWorks2019; Server=localhost";

            await using var connection = new SqlConnection(connectionString);
            connection.Open();

            await using var command = connection.CreateCommand();
            command.CommandText = "SELECT TOP 3 * FROM Production.Product";

            var data = Execute(command);

            var formattedData = data.First().ToTabularJsonString();

            this.Assent(formattedData.ToString(), _configuration);
        }

        private static IEnumerable<IEnumerable<IEnumerable<(string name, object value)>>> Execute(IDbCommand command)
        {
            using var reader = command.ExecuteReader();

            do
            {
                var values = new object[reader.FieldCount];
                var names = Enumerable.Range(0, reader.FieldCount).Select(reader.GetName).ToArray();

                // holds the result of a single statement within the query
                var table = new List<(string, object)[]>();

                while (reader.Read())
                {
                    reader.GetValues(values);
                    var row = new (string, object)[values.Length];
                    for (var i = 0; i < values.Length; i++)
                    {
                        row[i] = (names[i], values[i]);
                    }

                    table.Add(row);
                }

                yield return table;
            } while (reader.NextResult());
        }

        public void Dispose()
        {
            Formatter.ResetToDefault();
        }
    }
}