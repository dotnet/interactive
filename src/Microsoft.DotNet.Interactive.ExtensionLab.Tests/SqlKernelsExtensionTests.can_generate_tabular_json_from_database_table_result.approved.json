{
  "schema": {
    "primaryKey": [],
    "fields": [
      {
        "name": "ProductID",
        "type": "integer"
      },
      {
        "name": "Name",
        "type": "string"
      },
      {
        "name": "ProductNumber",
        "type": "string"
      },
      {
        "name": "MakeFlag",
        "type": "boolean"
      },
      {
        "name": "FinishedGoodsFlag",
        "type": "boolean"
      },
      {
        "name": "Color",
        "type": "any"
      },
      {
        "name": "SafetyStockLevel",
        "type": "any"
      },
      {
        "name": "ReorderPoint",
        "type": "any"
      },
      {
        "name": "StandardCost",
        "type": "number"
      },
      {
        "name": "ListPrice",
        "type": "number"
      },
      {
        "name": "Size",
        "type": "any"
      },
      {
        "name": "SizeUnitMeasureCode",
        "type": "any"
      },
      {
        "name": "WeightUnitMeasureCode",
        "type": "any"
      },
      {
        "name": "Weight",
        "type": "any"
      },
      {
        "name": "DaysToManufacture",
        "type": "integer"
      },
      {
        "name": "ProductLine",
        "type": "any"
      },
      {
        "name": "Class",
        "type": "any"
      },
      {
        "name": "Style",
        "type": "any"
      },
      {
        "name": "ProductSubcategoryID",
        "type": "any"
      },
      {
        "name": "ProductModelID",
        "type": "any"
      },
      {
        "name": "SellStartDate",
        "type": "datetime"
      },
      {
        "name": "SellEndDate",
        "type": "any"
      },
      {
        "name": "DiscontinuedDate",
        "type": "any"
      },
      {
        "name": "rowguid",
        "type": "any"
      },
      {
        "name": "ModifiedDate",
        "type": "datetime"
      }
    ]
  },
  "data": [
    {
      "ProductID": 1,
      "Name": "Adjustable Race",
      "ProductNumber": "AR-5381",
      "MakeFlag": false,
      "FinishedGoodsFlag": false,
      "Color": null,
      "SafetyStockLevel": 1000,
      "ReorderPoint": 750,
      "StandardCost": 0.0000,
      "ListPrice": 0.0000,
      "Size": null,
      "SizeUnitMeasureCode": null,
      "WeightUnitMeasureCode": null,
      "Weight": null,
      "DaysToManufacture": 0,
      "ProductLine": null,
      "Class": null,
      "Style": null,
      "ProductSubcategoryID": null,
      "ProductModelID": null,
      "SellStartDate": "2008-04-30T00:00:00",
      "SellEndDate": null,
      "DiscontinuedDate": null,
      "rowguid": "694215b7-08f7-4c0d-acb1-d734ba44c0c8",
      "ModifiedDate": "2014-02-08T10:01:36.827"
    },
    {
      "ProductID": 2,
      "Name": "Bearing Ball",
      "ProductNumber": "BA-8327",
      "MakeFlag": false,
      "FinishedGoodsFlag": false,
      "Color": null,
      "SafetyStockLevel": 1000,
      "ReorderPoint": 750,
      "StandardCost": 0.0000,
      "ListPrice": 0.0000,
      "Size": null,
      "SizeUnitMeasureCode": null,
      "WeightUnitMeasureCode": null,
      "Weight": null,
      "DaysToManufacture": 0,
      "ProductLine": null,
      "Class": null,
      "Style": null,
      "ProductSubcategoryID": null,
      "ProductModelID": null,
      "SellStartDate": "2008-04-30T00:00:00",
      "SellEndDate": null,
      "DiscontinuedDate": null,
      "rowguid": "58ae3c20-4f3a-4749-a7d4-d568806cc537",
      "ModifiedDate": "2014-02-08T10:01:36.827"
    },
    {
      "ProductID": 3,
      "Name": "BB Ball Bearing",
      "ProductNumber": "BE-2349",
      "MakeFlag": true,
      "FinishedGoodsFlag": false,
      "Color": null,
      "SafetyStockLevel": 800,
      "ReorderPoint": 600,
      "StandardCost": 0.0000,
      "ListPrice": 0.0000,
      "Size": null,
      "SizeUnitMeasureCode": null,
      "WeightUnitMeasureCode": null,
      "Weight": null,
      "DaysToManufacture": 1,
      "ProductLine": null,
      "Class": null,
      "Style": null,
      "ProductSubcategoryID": null,
      "ProductModelID": null,
      "SellStartDate": "2008-04-30T00:00:00",
      "SellEndDate": null,
      "DiscontinuedDate": null,
      "rowguid": "9c21aed2-5bfa-4f18-bcb8-f11638dc2e4e",
      "ModifiedDate": "2014-02-08T10:01:36.827"
    }
  ]
}